File: app/services/telegram/media_group_store.rb
class Telegram::MediaGroupStore
  def enqueue(message, wait_duration_seconds:)
    deadline_at = current_time_ms + wait_duration_ms(wait_duration_seconds)
    key = build_key(message)

    ApplicationRecord.transaction do
      MediaGroupBuffer.upsert(
        {
          key: key,
          deadline_at: deadline_at
        },
        unique_by: :index_media_group_buffers_on_key
      )

      MediaGroupMessage.upsert(
        {
          update_id: message.update_id,
          media_group_key: key,
          message_id: message.message_id,
          payload: JSON.generate(message.to_job_payload)
        },
        unique_by: :update_id
      )
    end

    { deadline_at: deadline_at, key: key }
  end

  def flush(key, expected_deadline_at:)
    ApplicationRecord.transaction do
      buffer = MediaGroupBuffer.lock.find_by(key: key)
      return { status: :missing } if buffer.nil?
      return { status: :stale } if buffer.deadline_at != expected_deadline_at

      wait_duration_ms = buffer.deadline_at - current_time_ms
      return { status: :pending, wait_duration_seconds: wait_duration_ms / 1000.0 } if wait_duration_ms.positive?

      rows = MediaGroupMessage.lock.where(media_group_key: key).order(:message_id, :update_id).to_a
      MediaGroupMessage.where(media_group_key: key).delete_all
      buffer.destroy!
      return { status: :missing } if rows.empty?

      { status: :ready, message: aggregate_messages(rows) }
    end
  end

  def clear!
    ApplicationRecord.transaction do
      MediaGroupMessage.delete_all
      MediaGroupBuffer.delete_all
    end
  end

  private

  def aggregate_messages(rows)
    messages = rows.map do |row|
      Telegram::InboundMessage.from_job_payload(JSON.parse(row.payload))
    end
    ordered_messages = messages.sort_by { |message| [message.message_id, message.update_id] }
    primary_message = ordered_messages.first
    aggregated_text = ordered_messages.filter_map { |message| message.text.presence }.first.to_s
    aggregated_image_file_ids = ordered_messages.flat_map(&:image_file_ids).uniq
    processing_updates = ordered_messages.map do |message|
      { update_id: message.update_id, message_id: message.message_id }
    end

    Telegram::InboundMessage.new(
      chat_id: primary_message.chat_id,
      image_file_ids: aggregated_image_file_ids,
      media_group_id: primary_message.media_group_id,
      message_id: primary_message.message_id,
      processing_updates: processing_updates,
      text: aggregated_text,
      user_id: primary_message.user_id,
      update_id: primary_message.update_id
    )
  end

  def build_key(message)
    "#{message.chat_id}:#{message.media_group_id}"
  end

  def current_time_ms
    (Time.now.to_f * 1000).to_i
  end

  def wait_duration_ms(wait_duration_seconds)
    (wait_duration_seconds * 1000).ceil
  end
end

File: src/main/java/com/telegramcodex/conversation/MediaGroupStore.java
package com.telegramcodex.conversation;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.telegramcodex.persistence.MediaGroupBufferEntity;
import com.telegramcodex.persistence.MediaGroupBufferJpaRepository;
import com.telegramcodex.persistence.MediaGroupMessageEntity;
import com.telegramcodex.persistence.MediaGroupMessageJpaRepository;
import com.telegramcodex.telegram.InboundMessage;
import jakarta.transaction.Transactional;
import java.util.Comparator;
import java.util.List;
import org.springframework.stereotype.Service;

@Service
public class MediaGroupStore {

    private final MediaGroupBufferJpaRepository bufferRepository;
    private final MediaGroupMessageJpaRepository messageRepository;
    private final ObjectMapper objectMapper;

    public MediaGroupStore(
        MediaGroupBufferJpaRepository bufferRepository,
        MediaGroupMessageJpaRepository messageRepository,
        ObjectMapper objectMapper
    ) {
        this.bufferRepository = bufferRepository;
        this.messageRepository = messageRepository;
        this.objectMapper = objectMapper;
    }

    @Transactional
    public EnqueueResult enqueue(InboundMessage message, double waitDurationSeconds) {
        long deadlineAt = System.currentTimeMillis() + Math.round(waitDurationSeconds * 1000.0);
        String key = buildKey(message);

        MediaGroupBufferEntity buffer = new MediaGroupBufferEntity();
        buffer.setKey(key);
        buffer.setDeadlineAt(deadlineAt);
        bufferRepository.save(buffer);

        MediaGroupMessageEntity row = new MediaGroupMessageEntity();
        row.setUpdateId(message.updateId());
        row.setMediaGroupKey(key);
        row.setMessageId(message.messageId());
        row.setPayload(writeMessage(message));
        messageRepository.save(row);

        return new EnqueueResult(deadlineAt, key);
    }

    @Transactional
    public FlushResult flush(String key, long expectedDeadlineAt) {
        MediaGroupBufferEntity buffer = bufferRepository.findById(key).orElse(null);
        if (buffer == null) {
            return FlushResult.missing();
        }
        if (buffer.getDeadlineAt() != expectedDeadlineAt) {
            return FlushResult.stale();
        }
        long waitDurationMs = buffer.getDeadlineAt() - System.currentTimeMillis();
        if (waitDurationMs > 0) {
            return FlushResult.pending(waitDurationMs / 1000.0);
        }
        List<MediaGroupMessageEntity> rows = messageRepository.findByMediaGroupKeyOrderByMessageIdAscUpdateIdAsc(key);
        messageRepository.deleteByMediaGroupKey(key);
        bufferRepository.delete(buffer);
        if (rows.isEmpty()) {
            return FlushResult.missing();
        }
        return FlushResult.ready(aggregateMessages(rows));
    }

    @Transactional
    public void clear() {
        messageRepository.deleteAll();
        bufferRepository.deleteAll();
    }

    private InboundMessage aggregateMessages(List<MediaGroupMessageEntity> rows) {
        List<InboundMessage> messages = rows.stream()
            .map(this::readMessage)
            .sorted(Comparator.comparingLong(InboundMessage::messageId).thenComparingLong(InboundMessage::updateId))
            .toList();
        InboundMessage primary = messages.getFirst();
        String aggregatedText = messages.stream()
            .map(InboundMessage::text)
            .filter(value -> value != null && !value.isBlank())
            .findFirst()
            .orElse(null);
        List<String> aggregatedImageFileIds = messages.stream()
            .flatMap(message -> message.imageFileIds().stream())
            .distinct()
            .toList();
        List<InboundMessage.ProcessingUpdate> processingUpdates = messages.stream()
            .map(message -> new InboundMessage.ProcessingUpdate(message.updateId(), message.messageId()))
            .toList();
        return new InboundMessage(
            primary.chatId(),
            aggregatedImageFileIds,
            primary.mediaGroupId(),
            primary.messageId(),
            null,
            processingUpdates,
            List.of(),
            null,
            null,
            null,
            null,
            null,
            aggregatedText,
            null,
            null,
            primary.userId(),
            primary.updateId()
        );
    }

    private String writeMessage(InboundMessage message) {
        try {
            return objectMapper.writeValueAsString(message);
        } catch (Exception error) {
            throw new IllegalStateException("Failed to persist media group payload", error);
        }
    }

    private InboundMessage readMessage(MediaGroupMessageEntity entity) {
        try {
            return objectMapper.readValue(entity.getPayload(), InboundMessage.class);
        } catch (Exception error) {
            throw new IllegalStateException("Failed to read media group payload", error);
        }
    }

    private String buildKey(InboundMessage message) {
        return message.chatId() + ":" + message.mediaGroupId();
    }

    public record EnqueueResult(long deadlineAt, String key) {
    }

    public record FlushResult(String status, InboundMessage message, Double waitDurationSeconds) {
        public static FlushResult missing() {
            return new FlushResult("missing", null, null);
        }

        public static FlushResult stale() {
            return new FlushResult("stale", null, null);
        }

        public static FlushResult pending(double waitDurationSeconds) {
            return new FlushResult("pending", null, waitDurationSeconds);
        }

        public static FlushResult ready(InboundMessage message) {
            return new FlushResult("ready", message, null);
        }
    }
}
